### PR TITLE
[CARBONDATA-2869] Add support for Avro Map data type support for SDK

### PR DESCRIPTION
**Pull request by manishgupta88**
This PR supports Avro Map data type support. It supports reading and writing complex map data type data using SDK.

For reading and writing Map data type using SDK refer the below test classes for reference
1. TestNonTransactionalCarbonTableForMapType.scala
2.  CarbonReaderTest.java

 - [ ] Any interfaces changed?
 No
 - [ ] Any backward compatibility impacted?
 No
 - [ ] Document update required?
No
 - [ ] Testing done
Add test cases  for verification
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
NA

### CODE REVIEW

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6330/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7982/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6705/

**Comment by jackylk** [comment]
Is this only for SDK or can be support in SQL also?

**Comment by jackylk** [comment]
Is map implemented as an array of struct<key, value>?

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6355/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8004/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6727/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6386/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6762/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6395/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8049/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6771/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6417/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8080/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/21/

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/TableSchemaBuilder.java [hunk 1]
```diff
@@ -232,6 +235,10 @@ private ColumnSchema addColumn(StructField field, String parentName, AtomicInteg
         for (int i = 0; i < fields.size(); i++) {
           addColumn(fields.get(i), parentFieldName, valIndex, false, true);
         }
+      } else if (field.getDataType().getName().equalsIgnoreCase("MAP")) {
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/metadata/schema/table/TableSchemaBuilder.java]: Please use `DataTypes.isMapType(field.getDataType())`

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/TableSchemaBuilder.java [hunk 2]
```diff
@@ -195,7 +196,8 @@ private ColumnSchema addColumn(StructField field, String parentName, AtomicInteg
     newColumn.setColumnReferenceId(newColumn.getColumnUniqueId());
     newColumn.setEncodingList(createEncoding(field.getDataType(), isSortColumn, isComplexChild));
     if (field.getDataType().isComplexType()) {
-      if (field.getDataType().getName().equalsIgnoreCase("ARRAY")) {
+      if (field.getDataType().getName().equalsIgnoreCase("ARRAY") || field.getDataType().getName()
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/metadata/schema/table/TableSchemaBuilder.java]: Please use `DataTypes.isArrayType(field.getDataType())`

File: core/src/main/java/org/apache/carbondata/core/scan/complextypes/MapQueryType.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * handles querying map data type columns
+ */
+public class MapQueryType extends ArrayQueryType {
+
+  public MapQueryType(String name, String parentname, int blockIndex) {
+    super(name, parentname, blockIndex);
+  }
+
+  @Override public Object getDataBasedOnDataType(ByteBuffer dataBuffer) {
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/scan/complextypes/MapQueryType.java]: Please add comment to tell how we convert array to map

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/AvroCarbonWriter.java [hunk 4]
```diff
@@ -134,6 +139,38 @@ private Object avroFieldToObject(Schema.Field avroField, Object fieldValue) {
         // also carbon internally needs float as double
         out = Double.parseDouble(fieldValue.toString());
         break;
+      case MAP:
+        // Note: Avro object takes care of removing the duplicates so we should not handle it again
+        // Map will be internally stored as Array<Struct<Key,Value>>
+        Map mapEntries = (HashMap) fieldValue;
+        Object[] arrayMapChildObjects = new Object[mapEntries.size()];
+        if (!mapEntries.isEmpty()) {
+          Iterator iterator = ((HashMap) fieldValue).entrySet().iterator();
```
Comment (ravipesala) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/AvroCarbonWriter.java]: Use `mapEntries` directly instead of typecast again

**Comment by manishgupta88** [comment]
@ravipesala ...fixed review comment. Kindly review and merge

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6451/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/66/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8137/

**Pull request closed without merging**